api/src/main/java/org/pmiops/workbench/initialcredits/InitialCreditsExpirationServiceImpl.java

```diff
@@ -17,4 +67,64 @@ public Optional<Timestamp> getCreditsExpiration(DbUser user) {
         // .filter(() -> institutionService.shouldBypassForCreditsExpiration(user))
         .map(DbUserInitialCreditsExpiration::getExpirationTime);
   }
+
+  private void checkExpiration(DbUser user, Timestamp now) {
+    DbUserInitialCreditsExpiration userInitialCreditsExpiration =
+        user.getUserInitialCreditsExpiration();
+    if (null != userInitialCreditsExpiration
+        && !userInitialCreditsExpiration.isBypassed()
+        && userInitialCreditsExpiration
+        .getNotificationStatus()
+        .equals(NotificationStatus.NO_NOTIFICATION_SENT)
+        && !(userInitialCreditsExpiration.getExpirationTime().after(now))) {
+      expireBillingStatusForUserWorkspaces(user);
+      deleteAppsAndRuntimesInFreeTierWorkspaces(user);
+      try {
+        mailService.alertUserInitialCreditsExpired(user);
+        userInitialCreditsExpiration.setNotificationStatus(
+            NotificationStatus.EXPIRATION_NOTIFICATION_SENT);
+        userDao.save(user);
+
+      } catch (MessagingException e) {
+        logger.error(
+            String.format(
+                "Failed to send initial credits expiration notification for user %s",
+                user.getUserId()));
+      }
+    }
+  }
+
+  private void expireBillingStatusForUserWorkspaces(DbUser user) {
+    workspaceDao.findAllByCreator(user).stream()
+        .filter(
+            ws ->
+                WorkspaceUtils.isFreeTier(
+                    ws.getBillingAccountName(), workbenchConfigProvider.get()))
```

Comment:
we should add two more filters in case we messed up something else or the other initialCredit cron job
1: workspace status is still active (not deleted) (we don't want to modify a already deleted state WS)
2: workspaceBilling status is `ACTIVE`

Reply:
Also, you are duplicating workspace lookup with `deleteAppsAndRuntimesInFreeTierWorkspaces` .
We can extract those workspaces to `deleteApp` and `expireBillingStatus` in upper class then pass workspace as parameter

Reply:
Good idea. I updated the code.